plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java

```diff
@@ -224,20 +227,92 @@ public void testMigrateUnsupportedColumnType()
+
+        assertQuery("SELECT * FROM " + hiveTableName, "SELECT 1");
+
+        assertUpdate("DROP TABLE " + hiveTableName);
+    }
+
+    @Test
+    public void testMigrateTimestampMicrosTypeWithAvro()
+    {
+        String tableName = "test_migrate_timestamp_micro_avro" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
+        Session session = timestampPrecisionSession(HiveTimestampPrecision.MICROSECONDS);
+
+        assertUpdate(session, "CREATE TABLE " + hiveTableName + " (x timestamp(6)) WITH (format='AVRO')");
+        assertUpdate(session, "INSERT INTO " + hiveTableName + " VALUES timestamp '2021-01-01 00:00:00.123456'", 1);
+        // Currently Hive maps timestamp(6) to timestamp(millis)
+        assertQuery("SELECT * FROM " + hiveTableName, "VALUES timestamp '2021-01-01 00:00:00.123'");
+
+        assertQueryFails("CALL iceberg.system.migrate('tpch', '" + tableName + "')", "Migrating timestamp\\(3\\) type with AVRO format is not supported.");
+
+        assertUpdate("DROP TABLE " + hiveTableName);
+    }
+
+    @Test
+    public void testMigrateTimestampMillisTypeWithAvro()
+    {
+        String tableName = "test_migrate_timestamp_millis_avro" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
 
+        assertUpdate("CREATE TABLE " + hiveTableName + " WITH (format='AVRO') AS SELECT timestamp '2021-01-01 00:00:00.000' x", 1);
         assertQuery("SELECT * FROM " + hiveTableName, "VALUES timestamp '2021-01-01 00:00:00.000'");
-        assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
+
+        assertQueryFails("CALL iceberg.system.migrate('tpch', '" + tableName + "')", "Migrating timestamp\\(3\\) type with AVRO format is not supported.");
 
         assertUpdate("DROP TABLE " + hiveTableName);
     }
 
+    @DataProvider
+    public static Object[][] fileFormatTimestampPrecisionsAndValues()
+    {
+        return new Object[][] {
+                {"ORC", HiveTimestampPrecision.MILLISECONDS, "2021-01-01 06:12:12.123", "2021-01-01 06:12:12.123"},
+                {"ORC", HiveTimestampPrecision.MICROSECONDS, "2021-01-01 06:12:12.123456", "2021-01-01 06:12:12.123456"},
+                {"ORC", HiveTimestampPrecision.NANOSECONDS, "2021-01-01 06:12:12.123456789", "2021-01-01 06:12:12.123457"}, // Truncated as Iceberg timestamp precision is 6
+                // 1.Hive saves timestamp without TZ in UTC and additionally adds writer_tz to Parquet. During read, Hiver takes timestamp written in UTC and adjust it according to writer_tz.
+                // Iceberg reads only timestamp in UTC, it does not adjust it according writer_tz value.
+                // 2.For the tests, Hive has a setting hive.parquet.time-zone="America/Bahia_Banderas"
+                {"PARQUET", HiveTimestampPrecision.MILLISECONDS, "2021-01-01 06:12:12.123", "2021-01-01 12:12:12.123"},
```

Comment:
This looks like a correctness problem, unless we can consider this a configuration problem.
What configuration would be needed to get back what was written?